### PR TITLE
Implement a conda dependency resolver.

### PR DESCRIPTION
**Pull request by jmchilton**
Much more information on planemo PR that I used to build a test bed for this:

https://github.com/galaxyproject/planemo/pull/392

Also syncs galaxy with planemo shared library changed.
- [x] Use tempfile (see commnt from @erasche).
- [x] Use conda exec on PATH (already in planemo, just sync it and test).
- [x] Update galaxy.ini.sample with description of conda options.
- [x] Update new documentation from @pvanheus to describe the dependency resolvers.

### CODE REVIEW

File: lib/galaxy/tools/deps/commands.py [hunk 1]
```diff
@@ -41,6 +74,16 @@ def __wait(cmds, **popen_kwds):
     return stdout
 
 
+def download_command(url, quote_url=False):
+    if quote_url:
+        url = "'%s'" % url
+    if which("wget"):
+        download_cmd = ["wget", "-q", "--recursive", "-O" "-", url]
+    else:
+        download_cmd = ["curl", url]
```
Comment (hexylena) [inline lib/galaxy/tools/deps/commands.py]: do we maybe want `curl -L` in case anyone has 301s/302s in the way?

File: lib/galaxy/tools/deps/conda_util.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        url = "https://repo.continuum.io/miniconda/Miniconda-latest-Linux-x86_64.sh"
+    return url
+
+
+def find_conda_prefix(conda_prefix=None):
+    """ If supplied conda_prefix is not set, default to the default location
+    for Miniconda installs.
+    """
+    if conda_prefix is None:
+        return os.path.join(os.path.expanduser("~"), "miniconda2")
```
Comment (bgruening) [inline lib/galaxy/tools/deps/conda_util.py:40]: I think it is safer to parse the output from `conda info` and if this is not successful we should fall back to `miniconda2`

File: lib/galaxy/tools/deps/conda_util.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+# BSD 3-clause
+CONDA_LICENSE = "http://docs.continuum.io/anaconda/eula"
+
+
+def conda_link():
+    if IS_OS_X:
+        url = "https://repo.continuum.io/miniconda/Miniconda-latest-MacOSX-x86_64.sh"
+    else:
+        url = "https://repo.continuum.io/miniconda/Miniconda-latest-Linux-x86_64.sh"
```
Comment (bgruening) [inline lib/galaxy/tools/deps/conda_util.py:31]: Should we mirror/ship a version with every Galaxy release?

Comment (hexylena) [inline lib/galaxy/tools/deps/conda_util.py:31]: +1, especially if it is completely self-contained and doesn't muck with anything outside of `$GALAXY_ROOT`

File: lib/galaxy/tools/deps/conda_util.py (hunk 2 above)
Comment (jmchilton) [inline lib/galaxy/tools/deps/conda_util.py:40]: This is a wonderful idea, I am on it.

File: lib/galaxy/tools/deps/conda_util.py (hunk 3 above)
Comment (jmchilton) [inline lib/galaxy/tools/deps/conda_util.py:31]: Galaxy doesn't ship with its dependencies like wheels so my instinct is probably not to do this. Maybe if someday there is an all-in-one deb sort of install it could make a lot of sense.

File: lib/galaxy/tools/deps/conda_util.py (hunk 2 above)
Comment (jmchilton) [inline lib/galaxy/tools/deps/conda_util.py:40]: Almost started writing an info parser, then I remembered `--json`. Planemo really needs a `--json` - it is amazing.

File: lib/galaxy/tools/deps/conda_util.py (hunk 3 above)
Comment (hexylena) [inline lib/galaxy/tools/deps/conda_util.py:31]: Can we stick it in a wheel? (and then post-download, make it accessible on the path?)

File: lib/galaxy/tools/deps/conda_util.py (hunk 2 above)
Comment (jmchilton) [inline lib/galaxy/tools/deps/conda_util.py:40]: Actually, I'll make this change for planemo - but I feel like Galaxy should by default manage its own conda install that doesn't touch `~` or use the existing one on the `PATH`.

File: lib/galaxy/tools/deps/conda_util.py (hunk 3 above)
Comment (bgruening) [inline lib/galaxy/tools/deps/conda_util.py:31]: We could download (and check before) during startup, like wheel fetching and than download a copy from cargo-store?

File: lib/galaxy/tools/deps/conda_util.py (hunk 2 above)
Comment (bgruening) [inline lib/galaxy/tools/deps/conda_util.py:40]: :+1: ! Would be good if we can reuse the `tool_deps` path from the current tool installations.

File: lib/galaxy/tools/deps/conda_util.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+            return {"channels": ["defaults"]}
+
+    def save_condarc(self, conf):
+        condarc = self.condarc
+        with open(condarc, "w") as f:
+            return yaml.safe_dump(conf, f)
+
+    @property
+    def condarc(self):
+        return os.path.join(os.path.expanduser("~"), ".condarc")
```
Comment (bgruening) [inline lib/galaxy/tools/deps/conda_util.py:105]: Same here @jmchilton. If Galaxy handles it's own installation we should look there.

File: lib/galaxy/tools/deps/conda_util.py (hunk 3 above)
Comment (hexylena) [inline lib/galaxy/tools/deps/conda_util.py:31]: @bgruening suggested wheels just to ensure homogenity of the dependency management.

File: lib/galaxy/tools/deps/conda_util.py (hunk 2 above)
Comment (jmchilton) [inline lib/galaxy/tools/deps/conda_util.py:40]: Noticed in JSON it is called default_prefix - should check if it is possible to use the conda on the PATH but override the location of the prefix.

Comment (jmchilton) [inline lib/galaxy/tools/deps/conda_util.py:40]: So I made the change in planemo, it will end up in Galaxy on the next sync after some more testing. Is this okay, or is it really important it be part of this PR. https://github.com/galaxyproject/planemo/commit/217eb9a5075bfa8a4127b6b8caff0556d4cbd516

File: lib/galaxy/tools/linters/inputs.py [hunk 5]
```diff
@@ -24,34 +24,50 @@ def lint_inputs(tool_xml, lint_ctx):
         if param_type == "data":
             if "format" not in param_attrib:
                 lint_ctx.warn("Param input [%s] with no format specified - 'data' format will be assumed.", param_name)
+
```
Comment (bgruening) [inline lib/galaxy/tools/linters/inputs.py:27]: This belongs in a different PR, isn't it?

Comment (jmchilton) [inline lib/galaxy/tools/linters/inputs.py:27]: No, this just synced all of planemo's libs. I think that is fine - let me know though.

File: lib/galaxy/tools/deps/resolvers/conda.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+from ..conda_util import (
+    CondaContext,
+    CondaTarget,
+    install_conda,
+    is_conda_target_installed,
+    install_conda_target,
+    build_isolated_environment,
+)
+
+DEFAULT_ENSURE_CHANNELS = "r,bioconda"
```
Comment (bgruening) [inline lib/galaxy/tools/deps/resolvers/conda.py:24]: This should be accessible via a config file to the admin.
For the big picture if would be nice for a tool to request channels and Galaxy will ask for confirmation.

Comment (hexylena) [inline lib/galaxy/tools/deps/resolvers/conda.py:24]: :+1:, that way the whole point of "you can prefer your own lab's X over devteam over IUC" will actually be really possible (if iuc/devteam stuff are separate to other bioconda/ who knows. The idea struck me and it seemed exciting)

File: lib/galaxy/tools/deps/conda_util.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    h = hashlib.new('sha256')
+    for conda_package in conda_packages:
+        h.update(conda_package.install_environment)
+    return h.hexdigest()
+
+
+def install_conda(conda_context=None):
+    conda_context = _ensure_conda_context(conda_context)
+    download_cmd = " ".join(commands.download_command(conda_link(), quote_url=True))
+    download_cmd = "%s > /tmp/conda.bash" % download_cmd
```
Comment (hexylena) [inline lib/galaxy/tools/deps/conda_util.py]: Probably should use `tempfile` here.

File: lib/galaxy/tools/deps/resolvers/conda.py (hunk 6 above)
Comment (jmchilton) [inline lib/galaxy/tools/deps/resolvers/conda.py:24]: It is configurable, if you follow the tracks I think conda_ensure_channels will cause the dependency resolver to pick these up. I guess I should update galaxy.ini.sample with these options.

File: lib/galaxy/tools/deps/conda_util.py (hunk 7 above)
Comment (jmchilton) [inline lib/galaxy/tools/deps/conda_util.py]: Yeah, I should. Okay, I guess I will have to revise this PR. I'll create a little TODO list.

File: lib/galaxy/tools/linters/inputs.py [hunk 8]
```diff
@@ -24,34 +24,50 @@ def lint_inputs(tool_xml, lint_ctx):
...
+        if param_type == "select":
+            select_name = param.get("name", None) or param.get("argument", None)
```
Comment (bgruening) [inline lib/galaxy/tools/linters/inputs.py]: param_attrib.get("argument") was used two lines above

File: lib/galaxy/tools/deps/resolvers/conda.py (hunk 6 above)
Comment (bgruening) [inline lib/galaxy/tools/deps/resolvers/conda.py:24]: Mh, haven't seen this yet. Yeah but please update the galaxy.ini :)
Thanks!

File: lib/galaxy/tools/linters/inputs.py (hunk 8 above)
Comment (jmchilton) [inline lib/galaxy/tools/linters/inputs.py]: Fixed on next sync with planemo https://github.com/galaxyproject/planemo/commit/89957f227ff9b369d3c340097170cdb5470adec6.

File: lib/galaxy/tools/deps/conda_util.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    install_cmd = "bash /tmp/conda.bash -b -p '%s'" % conda_context.conda_prefix
```
Comment (bgruening) [inline lib/galaxy/tools/deps/conda_util.py]: This one will return an error if the conda_prefix already exists. Depending on the decision about the conda_prefix (Galaxy only or not) we need to add `-f` here.

Comment (jmchilton) [inline lib/galaxy/tools/deps/conda_util.py]: We will never want to replace a conda right? If conda_prefix is configure to point to something that exists we should leave it alone and just use it.

**Comment by jmchilton** [comment]
I think I took care of the documentation issues discussed and made some other small modifications and bug fixes.

**Comment by jmchilton** [comment]
@bgruening Why WIP, I think this is ready for merging.

**Comment by bgruening** [comment]
@jmchilton as you can see this label was added before your last cleanup commit.

**Comment by jmchilton** [comment]
@bgruening Oppps, sorry didn't notice that.

**Comment by bgruening** [comment]
@jmchilton one test is failing and I jenkins is not working at the moment at least I can not the the reason why the test is failing.

**Comment by jmchilton** [comment]
That is on the list of transiently failing tests. I can't  get it to fail locally and I've tried once to guess a fix for Jenkins but it did not work. I'll open a PR later today to just remove the test.

**Comment by jmchilton** [comment]
https://github.com/galaxyproject/galaxy/pull/1401 Should remove that frustrating maxseconds.xml test.

**Comment by jmchilton** [comment]
Rebased the following patch into this commit: https://github.com/jmchilton/galaxy/commit/469ff93c3a37054a8713f8c3df7e8abaf57b501f.

This includes some fixes, some more utility functionality for reasoning about conda used in downstream branches, and a unit test verifying auto installation of conda and tool dependencies. Run the test with:

```
GALAXY_TEST_INCLUDE_SLOW=1  nosetests test/unit/tools/test_conda_resolution.py
```

File: lib/galaxy/tools/deps/conda_util.py (hunk 4 above)
Comment (hexylena) [inline lib/galaxy/tools/deps/conda_util.py:105]: Comments from IRC:

```
14:15:36   jmchilton | is conda on your path- I guess I should remove it from the
                     | path for that test
14:16:00     erasche | heh, yep, sure is
14:16:00   jmchilton | it will go to a local folder if conda isn't on your path    
14:16:07     erasche | that's... interesting. 
14:16:17     erasche | I wonder if that's something admins will need to be aware of
14:16:22     erasche | is there any reason to default to the system conda?
14:16:56   jmchilton | the admin gave it to galaxy so galaxy should use it ?
14:18:23     erasche | that feels galaxy centric. My view is  "The admin
                     |  installed conda at some point for some purpose, now Galaxy 
                     | is behaving in ways that might cause distributed filesystem
                     |  issues if HOME for the galaxy user isn't exported over the 
                     | network"
```

**Comment by hexylena** [comment]
Okay, the test passes for me normally, but fails for me with a locally installed conda outside of the context of Galaxy. I'm fine with merging this as long as we're aware of that going forward, since no one is using this code just yet.

```
Fetching packages ...
samtools-1.2-1 100% |#######################################################################################################| Time: 0:00:01   1.20 MB/s
Extracting packages ...
[      COMPLETE      ]|##########################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|##########################################################################################################################| 100%
#
# To activate this environment, use:
# $ source activate __package__samtools@__unversioned__
#
# To deactivate this environment, use:
# $ source deactivate
#
FAIL

======================================================================
FAIL: tools.test_conda_resolution.test_conda_resolution
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hxr/work/galaxy/.venv/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/hxr/work/galaxy/test/unit/tools/test_conda_resolution.py", line 37, in test_conda_resolution
    assert dependency.shell_commands(None) is not None
AssertionError

----------------------------------------------------------------------
Ran 1 test in 55.592s

FAILED (failures=1)
```
